[Core] Change version display behavior

Bockiii opened the pull request:
Hi,

this fixes #2434 .

In BridgeList view, a remoteVersion request is done to get either the latest commit sha or the latest release version (based on the currently used deployment method). A badge will be created if the installation is up-to-date or if an update is available.

Bockiii committed:
[Core] Fix prtester context issue

Bockiii committed:
[core] Display version tags

Bockiii committed:
design things

Bockiii committed:
Works like this

Bockiii committed:
remove pr tester changes

Bockiii committed:
linting

Bockiii committed:
more linting

Bockiii committed:
linting and var fix

Bockiii committed:
copy paste error

Bockiii committed:
fix var showing

Bockiii commented:
yaman gave me the solution for the var-display problem. Now it works as expected, but I would like a review @em92 @dvikan

dvikan commented:
This introduces a blocking http call which slows down the page load?
Can we do it in javascript with ajax instead?

Bockiii commented:
I just did some browser-load-time-tests. Same default whitelist.
Current page: ~68ms
With badge: ~90ms

If you're not planning to play counterstrike on the site, I would say those 22ms are not the worst.

As for javascript: Maybe, didnt try.

Bockiii commented:
Could this be a usecase for the load/savecachevalue ? An update check every 24 hours would probably be enough.

dvikan commented:
Yes. Ideally we use `getContents` or any similar http-function, instead of constructing a curl handle manually. You should also consider that sometimes github replies slower and sometimes they are down. That would completely hang the browser for a long time.

Bockiii committed:
switch to getcontents, implement caching

Bockiii commented:
So I switched to getContents and implemented the load/savecachevalue. For that, I had to change the functions from protected to public. Is there any downside to that?

Also, yaman debated that it should be an opt-in feature. I'm on the side of "no config, just show it" but could be swayed to say "opt-out". If this would be an opt-in, no one will ever see it since people who dont think about updating their installation proooobably dont look at optional config parameters long after the initial setup.

Bockiii committed:
linting

Bockiii commented:
With the cached value, load times are exactly the same as without the badge btw. I tested both again.
No badge: 44-68ms (varies because of network, host power etc)
With badge: First call 100-150ms (its annoying to test since you have to reset everything), second call 44-68ms

Bockiii committed:
fix php warning

yamanq commented:
This should be behind a configuration option, which I implemented [here](https://github.com/Bockiii/rss-bridge/pull/29). Some reasons that instance administrators may not want this behavior:
- Adds external resources on every load of the main page. If GitHub is down, I assume getContents will return an exception that prevents the main page from loading. Even if that is fixed, I don't like the idea of querying 2 external resources for little to no utility for me since I already follow RSS-Bridge.
- For those using git commits, it will almost always have an update available, making it useless
- Custom installations with commits not in master will be similarly affected

Bockiii commented:
> Adds external resources on every load of the main page. If GitHub is down, I assume getContents will return an exception that prevents the main page from loading. Even if that is fixed, I don't like the idea of querying 2 external resources for little to no utility for me since I already follow RSS-Bridge.

I would probably pack the request into a try-block and return the same value back as localversion, which would display the badge as "everythings okay" if it cant find github.com

> For those using git commits, it will almost always have an update available, making it useless

I actually raised that same point in the request :) . The question is: How else do you want to inform the user/admin that a new version is available?

Many applications put it into the settings page (with the about or something like that). The *arr suite does that for example. Other applications are very insistent on showing you that a new version is available (portainer is one of those that do it too much, wordpress is very pressing about updates, Plex has a constant "update available" icon in the top row, etc).

We dont have a settings page. We dont want popups, we dont want info banners at the top. So what else is there to do than a small badge at the bottom of the page (with the "about" block).

> Custom installations with commits not in master will be similarly affected

Correct. I dont think there are that many of those tbh. And I hope and assume that with the speed of development we are at currently, not many people see the need for a custom, patched instance. Most of the things are very configurable, for custom bridges you can either use the docker mechanism or just still do git pull. Worst comes to worse, change a file you dont want overwritten and add it to .gitignore. After the next update, the head should be the same again.

So again: I'm for just showing it (after making sure that a downside on shields.io or github wouldnt make the bridgelist not load), at max an opt-out (which would then fulfill yamans needs because it just wouldnt do anything). If we make it opt-in, it will never ever get used and we can just leave it out completely.

Bockiii commented:
An alternative to the current way (new function, call to github, using cache) would be the display of two badges. "Your Installed Version : " and "Current Version :" 

This would only require some string manipulation and two img src calls to shields.io. It would not have the "Update Available" information to it (as the two badges are not compared at all, just displayed).

Would that be better @yamanq ? It would be the same information as now (your current version) plus the info badge on what's current.

It should not cause any actual load time increase and if shields.io is down, the image just isnt displayed.

corenting commented:
Instead of making an opt-in or opt-out display option, maybe the update check should be a button, for example "Click to check if an update is available".
When the user clicks, it would replace the button with the version information.

It would prevent a blocking external request, and would also avoid doing an external request automatically on every load of the list page.

Bockiii commented:
> Instead of making an opt-in or opt-out display option, maybe the update check should be a button, for example "Click to check if an update is available". When the user clicks, it would replace the button with the version information.

I like that solution the least by far :D

> It would prevent a blocking external request, and would also avoid doing an external request automatically on every load of the list page.

I agree, thats why it could be in a try/catch block. I just didnt do it yet because the alternative would mean no call at all, only images. And if people prefer that, I dont need to work on the try/catch block :)

Bockiii commented:
So lets do a poll:

#### Part 1:
👀  : Current version (with try/catch to prevent stalling) with new function, opening up save/loadcache, use cache and a single badge
🚀 : Two badges (your installed version and remote latest version). No outside calls, only string manipulations and two images
👎 : I dont want this feature at all, keep it like it is

#### Part 2:
❤️ : No option
😕 : Opt-out (meaning configurable but default is active)
🎉 : Opt-in (configurable, default deactivated)

Choose one of each part.

yamanq commented:
I think the button idea is the best one so far. It has all the benefits of both including and not including the button. Also, for the shields.io badges, is there no way to download them and put them into `static/`?

For the two badges idea, wouldn't outside calls still be needed to fetch what the remote latest version is?

Bockiii commented:
> I think the button idea is the best one so far. It has all the benefits of both including and not including the button. Also, for the shields.io badges, is there no way to download them and put them into `static/`?

That would defeat the purpose of the badge. You would have to generate a new one for every commit and it wouldn't be possible to do that at all for the "current version"


> For the two badges idea, wouldn't outside calls still be needed to fetch what the remote latest version is?

No, it's shields io functionality. You just embed an img src and they do the processing. Check the ones we already have in the Readme or test them on their site. It wouldn't be a getcontent/curl/whatever call, just am img src

dvikan commented:
I did some testing over the period of 100 seconds. The average time was 142ms. The endpoint can be fetched from js because they allow any origin using CORS.

Bockiii commented:
> I did some testing over the period of 100 seconds. The average time was 142ms. The endpoint can be fetched from js because they allow any origin using CORS.

For which version? With caching? That sounds like your server is slow. How long was the load time without the code at all?

dvikan commented:
This was a standalone php script that made a curl request towards the github url that you used in your PR. It's useful to know the average time for the network call.

On my local environment the bridge list loads in about 55ms.

yamanq commented:
I misunderstood some things about shields.io. I think current behavior is probably the best way to go. To address the concerns of the linked issue, we can put in the issue template to update to the latest version before reporting bugs or something to that effect.

Bockiii commented:
So with 4 people participating, we have a draw between "dont do it at all" and "two badges" with opt-out.

I'll just close this then and the request as a "will not implement". I'm not a fan of the two badges and I can see that anyone using the latest tag on docker or just a git clone'd installation off gh will have a constant display of "youre not up to date".

So, closed.
